File: cogs/pactice.py
# cogs/practice.py — Admin-only /practice command for starting a bot duel
import os
import json
import asyncio
import aiohttp
import discord
from discord.ext import commands
from discord import app_commands

# ── REQUIRED RESTRICTIONS (from your message) ────────────────────────────────
ADMIN_ROLE_ID = 1173049392371085392
BATTLEFIELD_CHANNEL_ID = 1367986446232719484

# ── CONFIG: backend + UI endpoints ──────────────────────────────────────────
# Prefer environment variables; fall back to config.json if present.
# DUEL_BACKEND_URL should expose the /bot/practice route (e.g., https://your-app.onrailway.app)
# DUEL_UI_URL is your duel front-end page (will be opened by users)
def _load_config():
    cfg = {}
    try:
        with open("config.json", "r", encoding="utf-8") as f:
            cfg = json.load(f)
    except Exception:
        pass
    return cfg

_cfg = _load_config()
DUEL_BACKEND_URL = os.getenv("DUEL_BACKEND_URL", _cfg.get("duel_backend_base_url", "")).rstrip("/")
DUEL_UI_URL      = os.getenv("DUEL_UI_URL",      _cfg.get("duel_ui_url", "")).rstrip("/")

# Sensible defaults if nothing configured yet (prevents crashes, shows clear error)
if not DUEL_BACKEND_URL:
    DUEL_BACKEND_URL = "http://localhost:3000"  # adjust if your server uses a different port
if not DUEL_UI_URL:
    DUEL_UI_URL = "http://localhost:5173"       # adjust to your Duel UI host

PRACTICE_ENDPOINT = f"{DUEL_BACKEND_URL}/bot/practice"  # we added an alias to your backend

# ── VIEW WITH URL BUTTON ────────────────────────────────────────────────────
class OpenDuelUIButton(discord.ui.View):
    def __init__(self, url: str, timeout: float | None = 180):
        super().__init__(timeout=timeout)
        self.add_item(discord.ui.Button(label="Open Duel UI", url=url))

# ── COG ─────────────────────────────────────────────────────────────────────
class PracticeDuel(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot

    # Slash command: /practice
    @app_commands.command(
        name="practice",
        description="(Admin only) Start a practice duel vs the bot and get a private link to the Duel UI."
    )
    async def practice(self, interaction: discord.Interaction):
        # Channel restriction
        if interaction.channel_id != BATTLEFIELD_CHANNEL_ID:
            await interaction.response.send_message(
                "❌ This command can only be used in **#battlefield**.",
                ephemeral=True
            )
            return

        # Role restriction (Admin only)
        user_roles = getattr(interaction.user, "roles", [])
        if not any(getattr(r, "id", None) == ADMIN_ROLE_ID for r in user_roles):
            await interaction.response.send_message(
                "❌ You must have the **Admin** role to use this command.",
                ephemeral=True
            )
            return

        # Let the user know we’re starting the duel
        await interaction.response.defer(ephemeral=True, thinking=True)

        # Call backend to initialize a practice duel (build decks, draw 3, coin flip, etc.)
        try:
            async with aiohttp.ClientSession() as session:
                # 5s connect timeout / 15s read timeout
                timeout = aiohttp.ClientTimeout(total=20)
                async with session.get(PRACTICE_ENDPOINT, timeout=timeout) as resp:
                    if resp.status != 200:
                        text = await resp.text()
                        raise RuntimeError(f"Backend responded {resp.status}: {text[:300]}")
                    _ = await resp.json()  # duelState if you want to inspect/log

        except asyncio.TimeoutError:
            await interaction.edit_original_response(
                content="⚠️ The duel server timed out while starting practice. Try again in a moment."
            )
            return
        except Exception as e:
            await interaction.edit_original_response(
                content=f"⚠️ Failed to start practice duel:\n`{e}`\nCheck DUEL_BACKEND_URL or server logs."
            )
            return

        # Build the Duel UI link. You can append query params for your UI if desired.
        duel_url = f"{DUEL_UI_URL}?mode=practice"

        # Ephemeral confirmation + button to open the UI
        embed = discord.Embed(
            title="Practice Duel Ready",
            description=(
                "A fresh duel vs **Practice Bot** has been initialized.\n\n"
                "• Both sides start at **200 HP**\n"
                "• Each draws **3 cards**\n"
                "• **Coin flip** decides who goes first\n\n"
                "Click the button below to open the Duel UI."
            ),
            color=0x2ecc71
        )
        embed.set_footer(text="This message is visible only to you (ephemeral).")

        await interaction.edit_original_response(
            content=None,
            embed=embed,
            view=OpenDuelUIButton(duel_url)
        )

# ── SETUP ───────────────────────────────────────────────────────────────────
async def setup(bot: commands.Bot):
    await bot.add_cog(PracticeDuel(bot))

File: cogs/pactice.js
// commands/practice.js
import { SlashCommandBuilder, EmbedBuilder, ActionRowBuilder, ButtonBuilder, ButtonStyle, PermissionFlagsBits } from 'discord.js';

// ── CONFIG ───────────────────────────────────────────────────────────────────
// Prefer env vars; fall back to config.json if present; otherwise sensible local defaults.
import fs from 'fs';

const ADMIN_ROLE_ID = '1173049392371085392';
const BATTLEFIELD_CHANNEL_ID = '1367986446232719484';

let cfg = {};
try {
  const raw = fs.readFileSync('config.json', 'utf-8');
  cfg = JSON.parse(raw);
} catch (_) {}

const DUEL_BACKEND_URL = (process.env.DUEL_BACKEND_URL || cfg.duel_backend_base_url || 'http://localhost:3000').replace(/\/$/, '');
const DUEL_UI_URL      = (process.env.DUEL_UI_URL || cfg.duel_ui_url || 'http://localhost:5173').replace(/\/$/, '');

export const data = new SlashCommandBuilder()
  .setName('practice')
  .setDescription('(Admin only) Start a practice duel vs the bot and get a private link to open the Duel UI.')
  // no default member permissions here; we’ll enforce role + channel manually
  .setDMPermission(false);

export async function execute(interaction) {
  // Channel restriction
  if (interaction.channelId !== BATTLEFIELD_CHANNEL_ID) {
    await interaction.reply({ content: '❌ This command can only be used in **#battlefield**.', ephemeral: true });
    return;
  }

  // Role restriction (Admin only)
  const member = interaction.member; // GuildMember
  const hasAdminRole = member?.roles?.cache?.has(ADMIN_ROLE_ID);
  if (!hasAdminRole) {
    await interaction.reply({ content: '❌ You must have the **Admin** role to use this command.', ephemeral: true });
    return;
  }

  await interaction.deferReply({ ephemeral: true });

  // Call backend to initialize practice duel
  try {
    const res = await fetch(`${DUEL_BACKEND_URL}/bot/practice`, { method: 'GET' });
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`Backend responded ${res.status}: ${text.slice(0, 300)}`);
    }
    // const duelState = await res.json(); // not needed, but available for logging
  } catch (err) {
    await interaction.editReply({
      content: `⚠️ Failed to start practice duel:\n\`${String(err)}\`\nCheck DUEL_BACKEND_URL or your server logs.`
    });
    return;
  }

  const duelUrl = `${DUEL_UI_URL}?mode=practice`;

  const embed = new EmbedBuilder()
    .setTitle('Practice Duel Ready')
    .setDescription(
      [
        'A fresh duel vs **Practice Bot** has been initialized.',
        '',
        '• Both sides start at **200 HP**',
        '• Each draws **3 cards**',
        '• **Coin flip** decides who goes first',
        '',
        'Click the button below to open the Duel UI.'
      ].join('\n')
    )
    .setColor(0x2ecc71)
    .setFooter({ text: 'This message is visible only to you (ephemeral).' });

  const row = new ActionRowBuilder().addComponents(
    new ButtonBuilder()
      .setLabel('Open Duel UI')
      .setStyle(ButtonStyle.Link)
      .setURL(duelUrl)
  );

  await interaction.editReply({ embeds: [embed], components: [row] });
}
